src/invoice/mod.rs

```diff
@@ -130,8 +130,19 @@ impl Invoice {
         &mut self,
         signer_name: String,
         signer_role: SignatureRole,
-        key: Keypair,
+        key: &Keypair,
     ) -> Result<(), SignatureError> {
+        // The spec says it is illegal for the a single key to sign the same invoice
+        // more than once.
+        let encoded_key = base64::encode(key.public.to_bytes());
+        if let Some(sigs) = self.signature.as_ref() {
```

Comment:
This `if let` and `for` + `if` block could be reduced to something close to this:

```rust
if self.signature.as_ref().map(|sigs|  sigs.iter().any(|s| s.key == encoded_key) ).unwrap_or(false) {
    return Err(SignatureError::DuplicateSignature)
}
```

Reply:
This part is already re-written for the next PR... so gonna leave this for now and just merge it in the next PR